ocaml/xapi/xapi_pusb.ml

```diff
@@ -31,6 +32,67 @@ let create ~__context ~uSB_group ~host ~other_config ~path
   debug "PUSB ref='%s' created" (Ref.string_of pusb);
   pusb
 
+let scan_start ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let known_usb =
+    Db.PUSB.get_all_records ~__context
+    |> List.filter (fun (rf, rc) -> rc.API.pUSB_host = host)
+    |> get_known_usb
+  in
+  let local_usb =
+    get_usbs ~__context
+    |> get_local_usb
+  in
+  (* Create the newly added pusbs *)
+  USBSet.iter (fun s -> let self = create ~__context ~uSB_group:(Ref.null) ~host ~other_config:[] ~path:s.USB.path ~vendor_id:s.USB.vendor_id
+                   ~vendor_desc:s.USB.vendor_desc ~product_id:s.USB.product_id ~product_desc:s.USB.product_desc ~serial:s.USB.serial
+                   ~version:s.USB.version ~description:s.USB.description in
+               let group = Xapi_pusb_helpers.find_or_create ~__context self in
+               Db.PUSB.set_USB_group ~__context ~self ~value:group
+  ) (USBSet.diff local_usb known_usb);
+
+  List.filter (fun (rf, rc) -> USBSet.mem (extract_known_usb_info rc) (USBSet.diff known_usb local_usb)) (Db.PUSB.get_all_records ~__context)
+  |> List.iter (fun (self, _) ->
+                  let usb_group = Db.PUSB.get_USB_group ~__context ~self in
+                  Db.PUSB.destroy ~__context ~self;
+                  let vusbs = Db.USB_group.get_VUSBs ~__context ~self:usb_group in
+                  List.iter (fun vusb -> Db.VUSB.destroy ~__context ~self:vusb) vusbs;
+                  Db.USB_group.destroy ~__context ~self:usb_group)
+
+let thread_started : bool ref = ref false
+let task_processed : bool ref = ref false
 
 let scan ~__context =
-  debug "pusb scan to do"
+  (* This is for pusb-scan call which uevent will call or xapi restart. *)
+  let should_start_thread =
+    Mutex.execute mutex (
+      fun ()->
+        task_processed := false;
+        match !thread_started with
+        | false -> thread_started := true; true
+        | true -> false
+    ) in
+  (* If should_start_thread is true, then will create a thread to scan. *)
+  if should_start_thread then
+    ignore (Thread.create (
+              fun () ->
+                let rec run task_exist =
+                  match task_exist with
+                  | false -> ()
+                  | true ->
+                      let has_task = Mutex.execute mutex (
```

Comment:
What is the purpose of this logic? Why is it not enough to run a scanning thread that sleeps most of the time, scans, and updates xapi? What is the mutex protecting?

Reply:
I think a scanning thread that sleeps most of time ..... can do this. Here is another idea to do this. The mutex here is protect concurrently access the has_task.

Reply:
This looks very convoluted. Let me write some pseudocode as an example of what I was thinking:

```ocaml
let cond = Condition.create ()
let m = Mutex.create ()
let scan_required = ref false

open Stdext.Threadext

let execute_scan () =
  Printf.printf "Starting scan\n%!";
  Thread.delay 5.0;
  Printf.printf "Finished scan\n%!"

let start_thread exec =
  Thread.create (fun () ->
      while true do
        Mutex.execute m (fun () ->
            while not !scan_required do
              Condition.wait cond m
            done;

            (* scan_required is true now *)
            scan_required := false;
          );

        try
          exec ()
        with e ->
          ()
      done) ()

let notify () =
  Mutex.execute m (fun () ->
      scan_required := true;
      Condition.broadcast cond);
  Printf.printf "Notified\n%!"

let _ =
  start_thread execute_scan;
  notify ();
  Thread.delay 1.0;
  notify ();
  notify ();
  notify ();
  Thread.delay 20.0

```

Reply:
Note, the above code is also unit-testable since you can pass in an arbitrary function into start_thread (though don't use Thread.delay like I did!)